mp2-v1/src/query/batching_planner.rs

```diff
@@ -0,0 +1,458 @@
+            tree.build_subtree(num_leaves, None);
+        }
+
+        tree
+    }
+
+    /// Insert a node as a child nore of the node with key `parent_node_key`.
+    /// The node is inserted as root if `parent_node_key` is `None`
+    fn insert_as_child_of(
+        &mut self,
+        parent_node_key: Option<&UpdateTreeKey<ARITY>>,
+    ) -> UpdateTreeKey<ARITY> {
+        if let Some(parent_key) = parent_node_key {
+            // get parent node
+            let parent_node = self.nodes.get_mut(parent_key).expect(
+                format!(
+                    "Providing a non-existing parent key for insertion: {:?}",
+                    parent_key
+                )
+                .as_str(),
+            );
+            // get number of existing children for the parent node, which is needed to compute
+            // the key of the child to be inserted
+            let num_childrens = parent_node.children_keys.len();
+            let new_child_key = parent_key.children_key(num_childrens);
+            let child_node = ProvingTreeNode {
+                parent_key: Some(parent_key.clone()),
+                children_keys: vec![],
+            };
+            // insert new child in the set of children of the parent
+            parent_node.children_keys.push(new_child_key.clone());
+            assert!(
+                self.nodes
+                    .insert(new_child_key.clone(), child_node)
+                    .is_none(),
+                "Node with key {:?} already found in the tree",
+                new_child_key
+            );
+            new_child_key
+        } else {
+            // insert as root
+            let root = ProvingTreeNode {
+                parent_key: None,
+                children_keys: vec![],
+            };
+            let root_key = UpdateTreeKey((0, 0));
+            assert!(
+                self.nodes.insert(root_key.clone(), root).is_none(),
+                "Error: root node inserted multiple times"
+            );
+            root_key
+        }
+    }
+
+    /// Build a full subtree containing `num_leaves` leaf nodes. The subtree
+    /// is full since `num_leaves` is expected to be `ARITY^exp`, for `exp >= 0`.
+    /// `parent_node_key` is the key of the parent node of the root of the subtree
+    fn build_full_subtree(&mut self, num_leaves: usize, parent_node_key: &UpdateTreeKey<ARITY>) {
+        let root_key = self.insert_as_child_of(Some(&parent_node_key));
+        if num_leaves > 1 {
+            for _ in 0..ARITY {
+                self.build_full_subtree(num_leaves / ARITY, &root_key);
+            }
+        } else {
+            // it's a leaf node, so we add it to leaves
+            self.leaves.push(root_key);
+        }
+    }
+
+    /// Build a subtree containing `num_leaves` leaf nodes.
+    /// `parent_node_key` is the key of the parent node of the root of the subtree, if any
+    fn build_subtree(&mut self, num_leaves: usize, parent_node_key: Option<&UpdateTreeKey<ARITY>>) {
```

Comment:
If I understood correctly what you mean, this would be a simpler solution but will lead to a very unbalanced tree, which I think would be worse in our case since it means that we can parallelize less the aggregation proofs. Indeed, if we have a more balanced tree, then we can generate more proofs simultaneously, and so split better the proving workload in the network. The strategy I have implemented should allow to get a *reasonably* balanced tree (not sure it's the optimal one, but should be good enough), because we place as many nodes as possible inside *full* subtrees, which are balanced.